### ISSUE TITLE
Support custom encoder/decoder

### ISSUE THREAD

**Issue by bep** [asdine/storm#4]
Now it looks hardcoded to JSON:

https://github.com/asdine/storm/blob/master/save.go#L62

**Comment by asdine** [comment]
I believe the custom encoder/decoder should be stored in the `storm.DB` struct, that way the API remains simple and not be passed to `Save` or `Find`

A function type must be defined, something like this maybe:
```go
EncoderFunc func(v interface{}) ([]byte, error)
DecoderFunc func(data []byte, v interface{}) error
```

That way we can have:
```go
db.SetEncoder(json.Marshal)
db.SetDecoder(json.Unmarshal)
```

**Comment by bep** [comment]
As the decoder/encoder must be symmetric (or at least in any case I can think of -- it would lead to confusing bugs if you set the encoder, but forget the decoder or the other way around), it would make sense to:

Add a interface with a `Decode` and `Encode` function and implement the JSON case as the default one. This fits better with the other encoding packages out there (like gob).

The hardest part is to find a decent name for the interface ... `EncodeAndDecoder` ...?

**Comment by bep** [comment]
As I need this to try this ORM on my app, I can create a PR for this.

**Comment by asdine** [comment]
Sure!
For the name of the interface, we can mimic the go standard library: `Reader` + `Writer` = `ReadWriter`.
So what about `EncodeDecoder` ?

**Comment by bep** [comment]
>EncodeDecoder

That works.

**Comment by bep** [comment]
Tip: Just tested Storm with https://github.com/Sereal/Sereal/tree/master/Go -- with proper object refs support including circular references, works like a charm and is **way powerful**.

**Comment by asdine** [comment]
Do you think that's something that we should propose by default?

**Comment by bep** [comment]
> that we should propose by default?

Hmm... It is a little bit exotic, I just found a curious issue (see https://github.com/Sereal/Sereal/issues/130), but it is in some sense backed by booking.com so it is pretty serious. And it solves some really hard problems.

Go's JSON encoder goes into a stack overflow panic if you send a object circular ref into it. I created a bug in Go about it (I don't expect them to handle it in JSON, but a little bit friendlier). "Who would ever want to serialize object graphs with circular refs?" was a question from one in the Go team. Pretty common, if you ask me.

I would probably consider `gob` as a good default for Storm, but as the interfaces a are simple to implement I would suggest adding Sereal into package /storm/codec/sereal, or something so people only needs to "go get" it if they really want it.
